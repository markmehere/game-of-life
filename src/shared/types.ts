export type CellPoint = {
  c: number;
  r: number;
};

export type CellBoard = boolean[][];
