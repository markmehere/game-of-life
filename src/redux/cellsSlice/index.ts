import { cellsSlice } from "./cellsSlice";

export default cellsSlice.reducer;
