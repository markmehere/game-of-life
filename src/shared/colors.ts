export const color = {
  cellDead: "#dee6e9",
  cellAlive: "#1885e4",
};
