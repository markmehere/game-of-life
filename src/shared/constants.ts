export const BOARD_ROWS = 6;
export const BOARD_COLS = 6;
