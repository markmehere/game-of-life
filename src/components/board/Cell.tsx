import { useAppDispatch } from "redux/hooks";
import styled from "styled-components";
import { toggle } from "redux/cellsSlice/cellsSlice";
import { color } from "shared/colors";
import { BOARD_COLS } from "shared/constants";

interface ButtonStyleProps {
  alive: boolean;
}

const CellButton = styled.button<ButtonStyleProps>`
  background-color: ${(props) =>
    props.alive ? color.cellAlive : color.cellDead};
  border: 1px solid white;
  width: ${30 / BOARD_COLS}rem;
  height: ${30 / BOARD_COLS}rem;
`;

export interface CellProps {
  value: boolean;
  row: number;
  col: number;
}

export function Cell({ row, col, value }: CellProps) {
  const dispatch = useAppDispatch();

  return (
    <CellButton
      alive={value}
      aria-pressed={value}
      aria-label={`Cell ${row},${col}`}
      onClick={() => dispatch(toggle({ r: row, c: col }))}
    />
  );
}
