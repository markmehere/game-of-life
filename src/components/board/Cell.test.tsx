import { screen as s, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import { store } from "redux/store";
import { CellBoard } from "shared/types";
import { Cell } from "./Cell";
import * as hooks from "redux/hooks";

describe("<Cell />", () => {
  it("matches snapshot when dead", async () => {
    const wrapper = render(
      <Provider store={store}>
        <Cell value={false} row={1} col={2} />
      </Provider>
    );
    const cellButton = await s.findByRole("button");
    expect(cellButton).toHaveAttribute("aria-pressed", "false");
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("matches snapshot when alive", async () => {
    const wrapper = render(
      <Provider store={store}>
        <Cell value={true} row={1} col={2} />
      </Provider>
    );
    const cellButton = await s.findByRole("button");
    expect(cellButton).toHaveAttribute("aria-pressed", "true");
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("toggles correctly", async () => {
    const spy = jest.spyOn(hooks, "useAppDispatch");
    const mockDispatch = jest.fn();
    spy.mockReturnValue(mockDispatch);
    render(
      <Provider store={store}>
        <Cell value={false} row={1} col={2} />
      </Provider>
    );
    const cellButton = await s.findByRole("button");
    userEvent.click(cellButton);
    expect(mockDispatch.mock.calls[0][0].payload).toEqual({
      c: 2,
      r: 1,
    });
  });
});
