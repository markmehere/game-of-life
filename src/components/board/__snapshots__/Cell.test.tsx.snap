// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell /> matches snapshot when alive 1`] = `
<body>
  <div>
    <button
      aria-label="Cell 1,2"
      aria-pressed="true"
      class="sc-bdvvtL ipWMZu"
    />
  </div>
</body>
`;

exports[`<Cell /> matches snapshot when dead 1`] = `
<body>
  <div>
    <button
      aria-label="Cell 1,2"
      aria-pressed="false"
      class="sc-bdvvtL jrqIUZ"
    />
  </div>
</body>
`;
